modules/scalacheck/test/src-ce3/weaver/scalacheck/CheckersConcurrencyTest.scala

```diff
@@ -0,0 +1,98 @@
+package weaver
+package scalacheck
+
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicInteger
+
+import scala.concurrent.duration._
+
+import cats.effect.Outcome._
+import cats.effect.testkit.TestControl
+import cats.effect.{ IO, Ref }
+import cats.syntax.all._
+
+import weaver.TestStatus
+
+import org.scalacheck.Gen
+
+object CheckersConcurrencyTest extends SimpleIOSuite {
+  test("tests should wait for slower tests to succeed before completion") {
+
+    val minTests = 10
+    object CheckersConcurrencyTestNested extends SimpleIOSuite with Checkers {
+
+      override def checkConfig: CheckConfig =
+        super.checkConfig.copy(perPropertyParallelism = minTests * 5,
+                               minimumSuccessful = minTests,
+                               maximumDiscardRatio = 10)
+
+      private val atomicInt = new AtomicInteger(0)
+
+      loggedTest("nested") { log =>
+        forall(Gen.delay(Gen.const(atomicInt.incrementAndGet()))) { i =>
+          val sleepFor =
+            Duration.create(math.max(0L, minTests - i.toLong), TimeUnit.SECONDS)
+          IO.sleep(sleepFor) *> log.info(s"Ran gen $i").as(success)
+        }
+      }
+    }
+
+    val expectedLogs = (1 to minTests).map(i => s"Ran gen $i").toList
+    TestControl.execute(runSuite(CheckersConcurrencyTestNested))
+      .flatMap { control =>
+        control.advanceAndTick(1.second).replicateA(minTests) *> control.results
+      }
+      .map {
+        case Some(Succeeded(r)) =>
+          expect(r.count(_.status == TestStatus.Success) == 1) and
+            expect(r.forall(_.status == TestStatus.Success)) and
+            expect.same(
+              expectedLogs,
+              expectedLogs.intersect(r.flatMap(_.log.toList.map(_.msg)))
```

Comment:
not sure about this  `intersect` call there.

Reply:
because of parallelism there are additional runs of the generator (e.g. "Ran gen 11", "Ran gen 12"). These runs themselves are not a problem because the parallelism is speculative and works ahead. Using intersect was an easy way of filtering out the unneeded executions for the expecation

Reply:
Gotcha 👍